R/tostring.R

```diff
@@ -724,50 +724,51 @@ wrap_string <- function(str, width, collapse = NULL, smart = TRUE) {
+
   # Check if it went fine
   if (any(nchar(ret) > width)) {
     which_exceeded <- which(nchar(ret) > width)
 
-    if (smart) {
-      # Recursive for loop to take word interval
-      for (we_i in which_exceeded) {
-        # Is there space for some part of the next word?
-        char_threshold <- width * (2 / 3) + 0.01 # if too little space -> no previous word
-        smart_condition <- nchar(ret[we_i - 1]) + 1 < char_threshold # +1 is for spaces
-        if (we_i - 1 > 0 && smart_condition) {
-          we_interval <- unique(c(we_i - 1, we_i))
-          we_interval <- we_interval[
-            (we_interval < (length(ret) + 1)) &
-            (we_interval > 0)
-          ]
-        } else {
-          we_interval <- we_i
-        }
-        # Split words and collapse (needs unique afterwards)
-        ret[we_interval] <- split_words_by(
-          paste0(ret[we_interval], collapse = " "),
-          width
-        )
-        # Taking out repetitions if there are more than one
-        if (length(we_interval) > 1) {
-          ret <- ret[-we_interval[-1]]
-        }
-        # Paste together and rerun
-        ret <- paste0(ret, collapse = " ")
-        return(wrap_string(str = ret, width = width, collapse = collapse, smart = smart))
+    # Recursive for loop to take word interval
+    while (length(which_exceeded) > 0) {
+      we_i <- which_exceeded[1]
+      # Is there space for some part of the next word?
+      char_threshold <- width * (2 / 3) + 0.01 # if too little space -> no previous word
+      smart_condition <- nchar(ret[we_i - 1]) + 1 < char_threshold # +1 is for spaces
+      if (we_i - 1 > 0 && smart_condition) {
+        we_interval <- unique(c(we_i - 1, we_i))
+        we_interval <- we_interval[
+          (we_interval < (length(ret) + 1)) &
+          (we_interval > 0)
+        ]
+      } else {
+        we_interval <- we_i
+      }
+      # Split words and collapse (needs unique afterwards)
+      ret[we_interval] <- split_words_by(
+        paste0(ret[we_interval], collapse = " "),
+        width
+      )
+      # Taking out repetitions if there are more than one
+      if (length(we_interval) > 1) {
+        ret <- ret[-we_interval[-1]]
+        we_interval <- we_interval[1]
+      }
+      # Paste together and rerun if it is not the same as original ret
+      ret_collapse <- paste0(ret, collapse = " ")
+
+      # Checking if we are stuck in a loop
+      ori_wrapped_txt_v <- .go_stri_wrap(str, width)
+      cur_wrapped_txt_v <- .go_stri_wrap(ret_collapse, width)
+      if (!setequal(ori_wrapped_txt_v, cur_wrapped_txt_v)) {
+        return(wrap_string(str = ret_collapse, width = width, collapse = collapse))
+      } else {
+        # help function: Very rare case where the recursion is stuck in a loop
+        ret_tmp <- force_split_words_by(ret[we_interval], width) # here we_interval is only one ind
+        ret <- append(ret, ret_tmp, we_interval)[-we_interval]
+        which_exceeded <- which(nchar(ret) > width)
```

Comment:
force_split was necessary because `stringi` appears unable to split certain characters such as "x ." into smaller elements. I suspect that space and full stop are considered one word (even if normalize = TRUE) so it gets stuck in an eternal recursion, so we need to break it manually, append it and redefine the index of what is needed to be modified